client/packages/core/src/queryTypes.ts

```diff
@@ -438,6 +438,213 @@ type InstaQLOptions = {
+            ? {
+                $like?: string;
+              }
+            : {}) &
+            (R extends false
+              ? {
+                  $isNull?: boolean;
+                }
+              : {}) &
+            (I extends true
+              ? {
+                  $ilike?: string;
+                }
+              : {}))
+  : never;
+
+type ValidWhereValue<Input, AttrDef extends DataAttrDef<any, any, any>> =
+  AttrDef extends DataAttrDef<infer V, infer R, infer I>
+    ? Input extends V
+      ? V
+      : NonEmpty<WhereOperatorObject<Input, V, R, I>>
+    : never;
+
+type NoDistribute<T> = [T] extends [any] ? T : never;
+
+type ValidWhereObject<
+  Input extends { [key: string]: any },
+  Schema extends IContainEntitiesAndLinks<any, any>,
+  EntityName extends keyof Schema['entities'],
+> = keyof Input extends
+  | ValidDotPath<keyof Input, Schema, EntityName>
+  | keyof Schema['entities'][EntityName]['attrs']
+  | 'and'
+  | 'or'
+  | 'id'
+  ? {
+      [K in ValidDotPath<keyof Input, Schema, EntityName>]?: ValidWhereValue<
+        Input[K],
+        ExtractAttrFromDotPath<K, Schema, EntityName>
+      >;
+    } & {
+      [K in keyof Schema['entities'][EntityName]['attrs']]?: ValidWhereValue<
+        Input[K],
+        Schema['entities'][EntityName]['attrs'][K]
+      >;
+    } & {
+      and?: Input extends { and: Array<infer Item> }
+        ? ValidWhereObject<NoDistribute<Item>, Schema, EntityName>[]
+        : never;
+      or?: Input extends { or: Array<infer Item> }
+        ? ValidWhereObject<NoDistribute<Item>, Schema, EntityName>[]
+        : never;
+    } & {
+      // Special case for id
+      id?: ValidWhereValue<Input['id'], DataAttrDef<string, false, false>>;
+    }
+  : never;
+
+/**
+ * Extracts the attribute definition from a valid dot path.
+ * Given a dot path like "user.posts.title", this will resolve to the type of the "title" attribute.
+ * If the path is just a link name (e.g., "posts"), it defaults to the id field.
+ * Returns DataAttrDef or never
+ */
+type ExtractAttrFromDotPath<
+  Path extends string | number | symbol,
+  Schema extends IContainEntitiesAndLinks<any, any>,
+  EntityName extends keyof Schema['entities'],
+> = Path extends keyof Schema['entities'][EntityName]['attrs']
+  ? Schema['entities'][EntityName]['attrs'][Path]
+  : Path extends 'id'
+    ? DataAttrDef<string, false, false>
+    : Path extends `${infer LinkName}.${infer RestPath}`
+      ? LinkName extends keyof Schema['entities'][EntityName]['links']
+        ? ExtractAttrFromDotPath<
+            RestPath,
+            Schema,
+            Schema['entities'][EntityName]['links'][LinkName]['entityName']
+          >
+        : never
+      : Path extends keyof Schema['entities'][EntityName]['links']
+        ? DataAttrDef<string, false, false>
+        : never;
+
+type ValidQuery<
+  Q extends object,
+  S extends IContainEntitiesAndLinks<any, any>,
+> = S extends InstantUnknownSchemaDef
+  ? InstaQLParams<S>
```

Comment:
We can ship this version as is, but we may want to do some research about unifying things. It would suck to have to keep all the previous code to handle just this case.